Unable to import resource with a parent

<!--- Provide a general summary of the issue -->

I am trying to import a resource by specifying the `type`, `name`, `id`, `parent` and `provider` but it doesn't work.

## Expected Behavior
<!--- Tell us what should happen -->

I should be able to import the resource (or it should do nothing if it already exists).

## Current Behavior
<!--- Tell us what happens instead of the expected behavior -->

```
❯ pulumi import 'aws:iam/user:User' jason jason --parent 'jason=urn:pulumi:prod::root::foobar:auth:AdministratorUser::jason' --provider 'admin=urn:pulumi:prod::root::pulumi:providers:aws::admin::26c1d59e-b7f9-4542-99fe-2327ff869308'
Previewing import (prod)

View Live: https://app.pulumi.com/steinybot/root/prod/previews/abcdef

     Type                 Name       Plan
     pulumi:pulumi:Stack  root-prod

Resources:
    33 unchanged

Do you want to perform this import? yes
Importing (prod)

View Live: https://app.pulumi.com/steinybot/root/prod/updates/46

   Type  Name  Status

Outputs:
    ...

Resources:
    33 unchanged

Duration: 1s

error: no name for provider urn:pulumi:prod::root::foobar:auth:AdministratorUser::jason
```

I don't know what it means that there is "no name for provider".

I tried using a different name for the provider:
```
❯ pulumi import 'aws:iam/user:User' jason jason --parent 'jason=urn:pulumi:prod::root::foobar:auth:AdministratorUser::jason' --provider 'jason=urn:pulumi:prod::root::pulumi:providers:aws::admin::26c1d59e-b7f9-4542-99fe-2327ff869308'
Previewing import (prod)

View Live: https://app.pulumi.com/steinybot/root/prod/previews/abcdef

     Type                 Name       Plan       Info
     pulumi:pulumi:Stack  root-prod             1 error
 =   └─ aws:iam:User      jason      import     1 error

Diagnostics:
  pulumi:pulumi:Stack (root-prod):
    error: preview failed

  aws:iam:User (jason):
    error: Preview failed: unknown parent 'urn:pulumi:prod::root::pulumi:providers:aws::admin::26c1d59e-b7f9-4542-99fe-2327ff869308' for resource 'urn:pulumi:prod::root::pulumi:providers:aws$aws:iam/user:User::jason'
```

I don't know why it is now saying that the provider is an unknown parent. Seems like somehow the parent and provider are getting mixed up.

Just to see what would happen I tried reversing them:
```
❯ pulumi import 'aws:iam/user:User' jason jason --provider 'jason=urn:pulumi:prod::root::foobar:auth:AdministratorUser::jason' --parent 'admin=urn:pulumi:prod::root::pulumi:providers:aws::admin::26c1d59e-b7f9-4542-99fe-2327ff869308'
Previewing import (prod)

View Live: https://app.pulumi.com/steinybot/root/prod/previews/abcdef

     Type                 Name       Plan
     pulumi:pulumi:Stack  root-prod  running
panic: fatal: invalid module in type: expected 'pulumi:providers', got 'foobar:auth'

goroutine 79 [running]:
github.com/pulumi/pulumi/sdk/v2/go/common/util/contract.failfast(...)
	/private/tmp/pulumi-20201213-74568-1in5v9y/sdk/go/common/util/contract/failfast.go:23
github.com/pulumi/pulumi/sdk/v2/go/common/util/contract.AssertNoError(0x5be05a0, 0xc000dabf00)
	/private/tmp/pulumi-20201213-74568-1in5v9y/sdk/go/common/util/contract/assert.go:40 +0xe5
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*importer).registerProviders(0xc0001e5b00, 0x5c14c20, 0xc00067b0c0, 0xc00167ae80, 0x35, 0x100, 0x0)
	/private/tmp/pulumi-20201213-74568-1in5v9y/pkg/resource/deploy/import.go:188 +0xe3f
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*importer).importResources(0xc0001e5b00, 0x5c14c20, 0xc00067b0c0, 0xc000fdf540, 0xfc65a58)
	/private/tmp/pulumi-20201213-74568-1in5v9y/pkg/resource/deploy/import.go:286 +0x105
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*deploymentExecutor).importResources(0xc00000ff00, 0x5c14ce0, 0xc0001f6690, 0xfc65a58, 0xc000fc7400, 0x7fffffff, 0x0, 0x0, 0x0, 0x0, ...)
	/private/tmp/pulumi-20201213-74568-1in5v9y/pkg/resource/deploy/deployment_executor.go:450 +0x167
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*deploymentExecutor).Execute(0xc00000ff00, 0x5c14ce0, 0xc0001f6690, 0xfc65a58, 0xc000fc7400, 0x7fffffff, 0x0, 0x0, 0x0, 0x0, ...)
	/private/tmp/pulumi-20201213-74568-1in5v9y/pkg/resource/deploy/deployment_executor.go:138 +0x1247
github.com/pulumi/pulumi/pkg/v2/resource/deploy.(*Deployment).Execute(...)
	/private/tmp/pulumi-20201213-74568-1in5v9y/pkg/resource/deploy/deployment.go:383
github.com/pulumi/pulumi/pkg/v2/engine.(*deployment).run.func1(0x5c1f880, 0xc000fc7400, 0xc000fc7a40, 0x5c14ce0, 0xc0001f6690, 0xc0001f6601, 0xc0001cf900, 0xc00010f7a0)
	/private/tmp/pulumi-20201213-74568-1in5v9y/pkg/engine/deployment.go:251 +0x267
created by github.com/pulumi/pulumi/pkg/v2/engine.(*deployment).run
	/private/tmp/pulumi-20201213-74568-1in5v9y/pkg/engine/deployment.go:237 +0x306
```

## Steps to Reproduce
<!--- Provide a link to a live example, or an unambiguous set of steps to -->
<!--- reproduce this bug. Include code to reproduce, if relevant -->

```
import * as pulumi from "@pulumi/pulumi";
import {Input, Output} from "@pulumi/pulumi";
import * as aws from "@pulumi/aws";

const config = new pulumi.Config();

const accountID = config.require("account-id");

const provider = new aws.Provider("admin", {
    allowedAccountIds: [accountID],
    region: aws.config.requireRegion(),
});

const opts = { provider: provider, protect: true };

interface Tagged {
    readonly tags?: pulumi.Input<{[key: string]: pulumi.Input<string>}>;
}

function tagged<A extends Tagged>(a: A): A {
    const oldTags = a.tags ?? {};
    const newTags = {
        ...oldTags,
        "pulumi:Project": pulumi.getProject(),
        "pulumi:Stack": pulumi.getStack(),
    };
    return { ...a, tags: newTags };
}

const administrators = new aws.iam.Group("Administrators", {
    name: "Administrators"
}, opts);

new aws.iam.GroupPolicyAttachment("IAMFullAccess", {
    group: administrators.name,
    policyArn: "arn:aws:iam::aws:policy/IAMFullAccess",
}, opts);

new aws.iam.GroupPolicyAttachment("AWSOrganizationsFullAccess", {
    group: administrators.name,
    policyArn: "arn:aws:iam::aws:policy/AWSOrganizationsFullAccess",
}, opts);

class AdministratorUser extends pulumi.ComponentResource {
    constructor(name: string, opts?: pulumi.InvokeOptions) {
        super("foobar:auth:AdministratorUser", name, {}, opts);

        const childOpts = { ...opts, parent: this };

        const user = new aws.iam.User(name, {
            name: name
        }, childOpts);

        new aws.iam.UserGroupMembership("Administrator", {
            user: user.name,
            groups: [
                administrators.name
            ],
        }, childOpts);

        this.registerOutputs();
    }
}

new AdministratorUser("jason", opts)
```

The resource already exists in the stack state. Not sure if that is relevant but I would still expect the import to succeed (or do nothing):
```
{
                "urn": "urn:pulumi:prod::root::foobar:auth:AdministratorUser$aws:iam/user:User::jason",
                "custom": true,
                "id": "jason",
                "type": "aws:iam/user:User",
                "inputs": {
                    "__defaults": [
                        "forceDestroy",
                        "path"
                    ],
                    "forceDestroy": false,
                    "name": "jason",
                    "path": "/",
                    "tags": {}
                },
                "outputs": {
                    "arn": "arn:aws:iam::12345:user/jason",
                    "forceDestroy": false,
                    "id": "jason",
                    "name": "jason",
                    "path": "/",
                    "tags": {},
                    "uniqueId": "AIDASFACSFJBHVIOLWK26"
                },
                "parent": "urn:pulumi:prod::root::foobar:auth:AdministratorUser::jason",
                "protect": true,
                "provider": "urn:pulumi:prod::root::pulumi:providers:aws::admin::26c1d59e-b7f9-4542-99fe-2327ff869308",
                "propertyDependencies": {
                    "name": null
                }
            }
```

## Context (Environment)
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->
<!--- Provide a general summary of the issue in the Title above -->

--- a/pkg/cmd/pulumi/import.go
+++ b/pkg/cmd/pulumi/import.go
@@ -36,6 +36,7 @@ import (
 	"github.com/pulumi/pulumi/pkg/v3/codegen/schema"
 	"github.com/pulumi/pulumi/pkg/v3/engine"
 	"github.com/pulumi/pulumi/pkg/v3/resource/deploy"
+	"github.com/pulumi/pulumi/pkg/v3/resource/deploy/providers"
 	"github.com/pulumi/pulumi/pkg/v3/resource/stack"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
 	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/plugin"
@@ -59,12 +60,19 @@ func parseResourceSpec(spec string) (string, resource.URN, error) {
 		return "", "", fmt.Errorf("spec must be of the form name=URN")
 	}
 
-	name, urn := spec[:equals], spec[equals+1:]
+	name, urn := spec[:equals], resource.URN(spec[equals+1:])
 	if name == "" || urn == "" {
 		return "", "", fmt.Errorf("spec must be of the form name=URN")
 	}
 
-	return name, resource.URN(urn), nil
+	if !urn.IsValid() {
+		if ref, err := providers.ParseReference(string(urn)); err == nil {
+			return "", "", fmt.Errorf("expected a URN but got a Provider Reference, use '%s' instead", ref.URN())
+		}
+		return "", "", fmt.Errorf("expected a URN but got '%s'", urn)
+	}
+
+	return name, urn, nil
 }
 
 func makeImportFile(
@@ -73,7 +81,7 @@ func makeImportFile(
 	parentSpec, providerSpec, version string) (importFile, error) {
 
 	nameTable := map[string]resource.URN{}
-	resource := importSpec{
+	res := importSpec{
 		Type:       tokens.Type(typ),
 		Name:       tokens.QName(name),
 		ID:         resource.ID(id),
@@ -84,24 +92,32 @@ func makeImportFile(
 	if parentSpec != "" {
 		parentName, parentURN, err := parseResourceSpec(parentSpec)
 		if err != nil {
-			return importFile{}, fmt.Errorf("could not parse parent spec '%v': %w", parentSpec, err)
+			parentName = "parent"
+			parentURN = resource.URN(parentSpec)
+			if !parentURN.IsValid() {
+				return importFile{}, fmt.Errorf("invalid parent URN: '%s'", parentURN)
+			}
 		}
 		nameTable[parentName] = parentURN
-		resource.Parent = parentName
+		res.Parent = parentName
 	}
 
 	if providerSpec != "" {
 		providerName, providerURN, err := parseResourceSpec(providerSpec)
 		if err != nil {
-			return importFile{}, fmt.Errorf("could not parse provider spec '%v': %w", providerSpec, err)
+			providerName = "provider"
+			providerURN = resource.URN(providerSpec)
+		}
+		if _, exists := nameTable[providerName]; exists {
+			return importFile{}, fmt.Errorf("provider and parent must have distinct names, both were '%s'", providerName)
 		}
 		nameTable[providerName] = providerURN
-		resource.Provider = providerName
+		res.Provider = providerName
 	}
 
 	return importFile{
 		NameTable: nameTable,
-		Resources: []importSpec{resource},
+		Resources: []importSpec{res},
 	}, nil
 }
 
@@ -318,12 +334,25 @@ func newImportCmd() *cobra.Command {
 			"by default.\n" +
 			"\n" +
 			"Should you want to import your resource(s) without protection, you can pass\n" +
-			"`--protect=false` as an argument to the command. This will leave all resources unprotected." +
-			"\n" +
+			"`--protect=false` as an argument to the command. This will leave all resources unprotected.\n" +
 			"\n" +
 			"A single resource may be specified in the command line arguments or a set of\n" +
-			"resources may be specified by a JSON file. This file must contain an object\n" +
-			"of the following form:\n" +
+			"resources may be specified by a JSON file.\n" +
+			"\n" +
+			"If using the command line args directly, the type, name, id and optional flags\n" +
+			"must be provided.  For example:\n" +
+			"\n" +
+			"    pulumi import 'aws:iam/user:User' name id\n" +
+			"\n" +
+			"Or to fully specify parent and/or provider, subsitute the <urn> for each into the following:\n" +
+			"\n" +
+			"     pulumi import 'aws:iam/user:User' name id --parent 'parent=<urn>' --provider 'admin=<urn>'\n" +
+			"\n" +
+			"If using the JSON file format to define the imported resource(s), use this instead:\n" +
+			"\n" +
+			"     pulumi import -f import.json\n" +
+			"\n" +
+			"Where import.json is a file that matches the following JSON format:\n" +
 			"\n" +
 			"    {\n" +
 			"        \"nameTable\": {\n" +
@@ -360,7 +389,9 @@ func newImportCmd() *cobra.Command {
 			"specify a provider, it will be imported using the default provider for its type. A\n" +
 			"resource that does specify a provider may specify the version of the provider\n" +
 			"that will be used for its import.\n" +
+			"\n" +
 			"Each resource may specify which input properties to import with;\n" +
+			"\n" +
 			"If a resource does not specify any properties the default behaviour is to\n" +
 			"import using all required properties.\n",
 		Run: cmdutil.RunResultFunc(func(cmd *cobra.Command, args []string) result.Result {

--- a/pkg/codegen/importer/hcl2.go
+++ b/pkg/codegen/importer/hcl2.go
@@ -125,7 +125,7 @@ func makeResourceOptions(state *resource.State, names NameTable) (*model.Block, 
 		if !providers.IsDefaultProvider(ref.URN()) {
 			name, ok := names[ref.URN()]
 			if !ok {
-				return nil, fmt.Errorf("no name for provider %v", state.Parent)
+				return nil, fmt.Errorf("no name for provider %v", state.Provider)
 			}
 			resourceOptions = appendResourceOption(resourceOptions, "provider", newVariableReference(name))
 		}

